crates/router/src/utils.rs

```diff
@@ -129,13 +129,49 @@ impl<E> ConnectorResponseExt
     }
 }
 
+pub fn is_zero_decimal_currency(currency: diesel_models::enums::Currency) -> bool {
+    matches!(
+        currency,
+        diesel_models::enums::Currency::BIF
+            | diesel_models::enums::Currency::CLP
+            | diesel_models::enums::Currency::DJF
+            | diesel_models::enums::Currency::GNF
+            | diesel_models::enums::Currency::JPY
+            | diesel_models::enums::Currency::KMF
+            | diesel_models::enums::Currency::KRW
+            | diesel_models::enums::Currency::MGA
+            | diesel_models::enums::Currency::PYG
+            | diesel_models::enums::Currency::RWF
+            | diesel_models::enums::Currency::UGX
+            | diesel_models::enums::Currency::VND
+            | diesel_models::enums::Currency::VUV
+            | diesel_models::enums::Currency::XAF
+            | diesel_models::enums::Currency::XOF
+            | diesel_models::enums::Currency::XPF
+    )
+}
+
+pub fn is_three_decimal_currency(currency: diesel_models::enums::Currency) -> bool {
+    matches!(
+        currency,
+        diesel_models::enums::Currency::BHD
+            | diesel_models::enums::Currency::JOD
+            | diesel_models::enums::Currency::KWD
+            | diesel_models::enums::Currency::OMR
+    )
+}
+
 /// Convert the amount to its base denomination based on Currency and return String
 pub fn to_currency_base_unit(
     amount: i64,
     currency: diesel_models::enums::Currency,
 ) -> Result<String, error_stack::Report<errors::ValidationError>> {
     let amount_f64 = to_currency_base_unit_asf64(amount, currency)?;
-    Ok(format!("{amount_f64:.2}"))
+    if is_zero_decimal_currency(currency) {
+        Ok(amount_f64.to_string())
+    } else {
+        Ok(format!("{amount_f64:.2}"))
+    }
```

Comment:
This could break existing flows as this is being used in connector utils and Gpay session calls. Let's not take this change in this PR. If you need decimal checks for String conversion you can use `to_currency_base_unit_with_zero_decimal_check()` method